### PR TITLE
8314125: RISC-V: implement Base64 intrinsic - encoding

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review the patch?

I'm also working a base64 decode instrinsic, but there is some performance regression in some cases, and decode and encode are totally independent with each other, so I will send out review of decode in another pr when I fix the performance regression in it.

Thanks.

## Test
benchmarks run on CanVM-K230 (vlenb == 16), and banana-pi (vlenb == 32)

I've tried several implementations, respectively with vector group
* m2+m1+scalar
* m2+scalar
* m1+scalar
* pure scalar
The best one is combination of m2+m1, it have best performance in all source size.

### K230

this implementation (m2+m1)
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark | (maxNumBytes) | Mode | Cnt | Score -intrinsic | Score + instrinsic, m1+m2 | Error | Units | -intrinsic/+intrinsic
-- | -- | -- | -- | -- | -- | -- | -- | --
Base64Encode.testBase64Encode | 1 | avgt | 10 | 86.784 | 86.996 | 0.459 | ns/op | 0.9975631063
Base64Encode.testBase64Encode | 2 | avgt | 10 | 93.603 | 94.026 | 1.081 | ns/op | 0.9955012443
Base64Encode.testBase64Encode | 3 | avgt | 10 | 121.927 | 123.227 | 0.342 | ns/op | 0.989450364
Base64Encode.testBase64Encode | 6 | avgt | 10 | 139.554 | 137.4 | 1.221 | ns/op | 1.015676856
Base64Encode.testBase64Encode | 7 | avgt | 10 | 160.698 | 162.25 | 2.36 | ns/op | 0.9904345146
Base64Encode.testBase64Encode | 9 | avgt | 10 | 161.085 | 153.772 | 1.505 | ns/op | 1.047557423
Base64Encode.testBase64Encode | 10 | avgt | 10 | 187.963 | 174.763 | 1.204 | ns/op | 1.075530862
Base64Encode.testBase64Encode | 48 | avgt | 10 | 405.212 | 199.4 | 6.374 | ns/op | 2.032156469
Base64Encode.testBase64Encode | 512 | avgt | 10 | 3652.555 | 1111.009 | 3.462 | ns/op | 3.287601631
Base64Encode.testBase64Encode | 1000 | avgt | 10 | 7217.187 | 2011.943 | 227.784 | ns/op | 3.587172698
Base64Encode.testBase64Encode | 20000 | avgt | 10 | 135165.706 | 33864.592 | 57.557 | ns/op | 3.991357876

</google-sheets-html-origin>

vector with only m2
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark | (maxNumBytes) | Mode | Cnt | Score -intrinsics | Score +intrinsic, m2 | Error | Units | -intrinsic/+intrinsic
-- | -- | -- | -- | -- | -- | -- | -- | --
Base64Encode.testBase64Encode | 1 | avgt | 10 | 86.797 | 86.872 | 0.374 | ns/op | 0.9991366608
Base64Encode.testBase64Encode | 2 | avgt | 10 | 93.971 | 94.203 | 1.918 | ns/op | 0.9975372334
Base64Encode.testBase64Encode | 3 | avgt | 10 | 122.074 | 123.978 | 1.009 | ns/op | 0.9846424366
Base64Encode.testBase64Encode | 6 | avgt | 10 | 138.999 | 138.344 | 2.175 | ns/op | 1.004734575
Base64Encode.testBase64Encode | 7 | avgt | 10 | 160.857 | 157.494 | 1.036 | ns/op | 1.021353194
Base64Encode.testBase64Encode | 9 | avgt | 10 | 161.511 | 154.998 | 1.727 | ns/op | 1.042019897
Base64Encode.testBase64Encode | 10 | avgt | 10 | 186.228 | 175.38 | 0.62 | ns/op | 1.061854259
Base64Encode.testBase64Encode | 48 | avgt | 10 | 408.461 | 349.558 | 15.377 | ns/op | 1.168507086
Base64Encode.testBase64Encode | 512 | avgt | 10 | 3679.283 | 1103.717 | 3.911 | ns/op | 3.333538398
Base64Encode.testBase64Encode | 1000 | avgt | 10 | 7206.265 | 1988.927 | 224.732 | ns/op | 3.623192304
Base64Encode.testBase64Encode | 20000 | avgt | 10 | 135695.875 | 33930.292 | 97.85 | ns/op | 3.99925456

</google-sheets-html-origin>

vector with only m1
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark | (maxNumBytes) | Mode | Cnt | Score -intrinsic | Score +intrinsic, m1 | Error | Units | -intrinsic/+intrinsic
-- | -- | -- | -- | -- | -- | -- | -- | --
Base64Encode.testBase64Encode | 1 | avgt | 10 | 86.837 | 87.137 | 0.527 | ns/op | 0.9965571456
Base64Encode.testBase64Encode | 2 | avgt | 10 | 94.723 | 94.125 | 5.122 | ns/op | 1.006353254
Base64Encode.testBase64Encode | 3 | avgt | 10 | 121.51 | 123.082 | 0.854 | ns/op | 0.9872280268
Base64Encode.testBase64Encode | 6 | avgt | 10 | 139.045 | 137.175 | 0.201 | ns/op | 1.013632222
Base64Encode.testBase64Encode | 7 | avgt | 10 | 161.216 | 159.387 | 2.385 | ns/op | 1.011475214
Base64Encode.testBase64Encode | 9 | avgt | 10 | 160.541 | 154.19 | 1.665 | ns/op | 1.041189442
Base64Encode.testBase64Encode | 10 | avgt | 10 | 184.874 | 174.766 | 5.569 | ns/op | 1.057837337
Base64Encode.testBase64Encode | 48 | avgt | 10 | 405.124 | 199.333 | 1.584 | ns/op | 2.032398047
Base64Encode.testBase64Encode | 512 | avgt | 10 | 3659.335 | 1185.626 | 24.686 | ns/op | 3.086415952
Base64Encode.testBase64Encode | 1000 | avgt | 10 | 7239.269 | 2164.709 | 1022.367 | ns/op | 3.344222711
Base64Encode.testBase64Encode | 20000 | avgt | 10 | 135048.828 | 38248.645 | 319.978 | ns/op | 3.530813392

</google-sheets-html-origin>

### banana-pi
data
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark banana-pi (vlenb==32) | (maxNumBytes) | Mode | Cnt | Score +intrinsic+rvv | Score +intrinsic-rvv | Score -intrinsic+rvv | Error | Units | Improvement +rvv | Improvement -rvv (scalar)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Base64Encode.testBase64Encode | 1 | avgt | 10 | 81.855 | 81.823 | 81.716 | 0.059 | ns/op | 0.998 | 0.999
Base64Encode.testBase64Encode | 2 | avgt | 10 | 87.864 | 88.043 | 87.932 | 0.048 | ns/op | 1.001 | 0.999
Base64Encode.testBase64Encode | 3 | avgt | 10 | 101.145 | 99.844 | 102.414 | 0.201 | ns/op | 1.013 | 1.026
Base64Encode.testBase64Encode | 6 | avgt | 10 | 122.91 | 122.863 | 128.995 | 0.122 | ns/op | 1.05 | 1.05
Base64Encode.testBase64Encode | 7 | avgt | 10 | 129.115 | 125.626 | 135.366 | 0.794 | ns/op | 1.048 | 1.078
Base64Encode.testBase64Encode | 9 | avgt | 10 | 136.16 | 131.854 | 143.428 | 0.215 | ns/op | 1.053 | 1.088
Base64Encode.testBase64Encode | 10 | avgt | 10 | 141.178 | 137.82 | 149.188 | 0.776 | ns/op | 1.057 | 1.082
Base64Encode.testBase64Encode | 48 | avgt | 10 | 234.022 | 232.435 | 273.815 | 2.816 | ns/op | 1.17 | 1.178
Base64Encode.testBase64Encode | 512 | avgt | 10 | 598.189 | 1249.987 | 1604.611 | 0.463 | ns/op | 2.682 | 1.284
Base64Encode.testBase64Encode | 1000 | avgt | 10 | 906.706 | 2350.467 | 3000.369 | 0.164 | ns/op | 3.309 | 1.276
Base64Encode.testBase64Encode | 20000 | avgt | 10 | 19222.596 | 43884.376 | 60557.76 | 26.96 | ns/op | 3.15 | 1.38

</google-sheets-html-origin>

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314125](https://bugs.openjdk.org/browse/JDK-8314125): RISC-V: implement Base64 intrinsic - encoding (**Enhancement** - P4)

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) 🔄 Re-review required (review applies to [736f5f8b](https://git.openjdk.org/jdk/pull/19973/files/736f5f8b449d39405846966b9931fe6493e5f62b))
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19973/head:pull/19973` \
`$ git checkout pull/19973`

Update a local copy of the PR: \
`$ git checkout pull/19973` \
`$ git pull https://git.openjdk.org/jdk.git pull/19973/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19973`

View PR using the GUI difftool: \
`$ git pr show -t 19973`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19973.diff">https://git.openjdk.org/jdk/pull/19973.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19973#issuecomment-2200275022)

### CODE REVIEW

**Comment by Hamlin-Li** [comment]
with pure scalar impelmentation, it also bring some performance imrpovement in all source size, so also enable the intrinsic when rvv is not supported.

performance data
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark | (maxNumBytes) | Mode | Cnt | Score -intrinsic | Score +instrinsic, scalar | Error | Units | Perf opt
-- | -- | -- | -- | -- | -- | -- | -- | --
Base64Encode.testBase64Encode | 1 | avgt | 10 | 86.784 | 86.75 | 0.38 | ns/op | 1
Base64Encode.testBase64Encode | 2 | avgt | 10 | 93.71 | 93.824 | 1.954 | ns/op | 0.999
Base64Encode.testBase64Encode | 3 | avgt | 10 | 121.824 | 123.487 | 0.559 | ns/op | 0.987
Base64Encode.testBase64Encode | 6 | avgt | 10 | 138.984 | 137.697 | 0.273 | ns/op | 1.009
Base64Encode.testBase64Encode | 7 | avgt | 10 | 161.243 | 157.696 | 0.875 | ns/op | 1.022
Base64Encode.testBase64Encode | 9 | avgt | 10 | 169.724 | 155.223 | 1.908 | ns/op | 1.093
Base64Encode.testBase64Encode | 10 | avgt | 10 | 185.92 | 176.339 | 5.875 | ns/op | 1.054
Base64Encode.testBase64Encode | 48 | avgt | 10 | 408.467 | 347.269 | 1.799 | ns/op | 1.176
Base64Encode.testBase64Encode | 512 | avgt | 10 | 3665.34 | 2718.442 | 26.954 | ns/op | 1.348
Base64Encode.testBase64Encode | 1000 | avgt | 10 | 7022.025 | 5290.003 | 33.216 | ns/op | 1.327
Base64Encode.testBase64Encode | 20000 | avgt | 10 | 135819.7 | 101988.94 | 2209.887 | ns/op | 1.332

</google-sheets-html-origin>

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -1820,6 +1820,17 @@ enum Nf {
...
+#define INSN(NAME, op, width, umop, mop, mew, nf)                                               \
+  void NAME(VectorRegister Vd_or_Vs3, Register Rs1, VectorMask vm = unmasked) { \
+    patch_VLdSt(op, Vd_or_Vs3, width, Rs1, umop, vm, mop, mew, nf);                         \
+  }
+
+  // Vector Unit-Stride Instructions
+  INSN(vlseg3e8_v, 0b0000111, 0b000, 0b00000, 0b00, 0b0, g3);
+
+  INSN(vsseg4e8_v, 0b0100111, 0b000, 0b00000, 0b00, 0b0, g4);
+#undef INSN
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:1832]: Blank like before the `#undef INSN`

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 2]
```diff
@@ -1820,6 +1820,17 @@ enum Nf {
 
 #undef INSN
 
+#define INSN(NAME, op, width, umop, mop, mew, nf)                                               \
+  void NAME(VectorRegister Vd_or_Vs3, Register Rs1, VectorMask vm = unmasked) { \
+    patch_VLdSt(op, Vd_or_Vs3, width, Rs1, umop, vm, mop, mew, nf);                         \
+  }
+
+  // Vector Unit-Stride Instructions
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:1828]:
```suggestion
  // Vector Unit-Stride Load Instructions
```

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 3]
```diff
@@ -1820,6 +1820,17 @@ enum Nf {
...
+  INSN(vlseg3e8_v, 0b0000111, 0b000, 0b00000, 0b00, 0b0, g3);
+
+  INSN(vsseg4e8_v, 0b0100111, 0b000, 0b00000, 0b00, 0b0, g4);
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:1831]:
```suggestion
  // Vector Unit-Stride Store Instructions
  INSN(vsseg4e8_v, 0b0100111, 0b000, 0b00000, 0b00, 0b0, g4);
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 4]
```diff
@@ -5103,6 +5103,228 @@ class StubGenerator: public StubCodeGenerator {
...
+
+  /**
+   * vector registers:
+   *   input VectorRegister's:  intputV1-V3, for m2 they could be v2, v4, v6, for m1 they could be v1, v2, v3
+   *   index VectorRegister's:  idxV1-V4, for m2 they could be v8, v10, v12, v14, for m1 they could be v4, v5, v6, v7
+   *   output VectorRegister's: outputV1-V4, for m2 they could be v16, v18, v20, v22, for m1 they could be v8, v9, v10, v11
+   *
+   * NOTE: each field will occupy a vector register group
+   */
+  void encodeVector(Register src, Register dst, Register codec, Register step,
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5115]:
```suggestion
  void generate_base64_encodeVector(Register src, Register dst, Register codec, Register step,
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 5]
```diff
@@ -5103,6 +5103,228 @@ class StubGenerator: public StubCodeGenerator {
...
+    __ BIND(ProcessData);
+
+    Label ProcessScalar, ProcessM2, ProcessM1;
+
+    if (!UseRVV) {
+      __ j(ProcessScalar);
+    }
+
+    // vector version
+    {
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5230]: You should not even generate the vectorized code if `UseRVV` is false. You can then remove https://github.com/openjdk/jdk/pull/19973/files#diff-97f199af6d1c8c17b2fa4f50eb1bbc0081858cc59a899f32792a2d31f933ccc4R5225-R5227 

```suggestion
    if (UseRVV) {
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 6]
```diff
@@ -5103,6 +5103,228 @@ class StubGenerator: public StubCodeGenerator {
...
+      encodeVector(src, dst, codec, limitM1,
+                    v1, v2, v3,         // inputs
+                    v4, v5, v6, v7,     // indexes
+                    v8, v9, v10, v11,   // outputs
+                    Assembler::m1);
+      __ sub(length, length, limitM1);
+    }
+
+    // scalar version
+    __ BIND(ProcessScalar);
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5263]: You can move that in the previous block at https://github.com/openjdk/jdk/pull/19973/files#diff-97f199af6d1c8c17b2fa4f50eb1bbc0081858cc59a899f32792a2d31f933ccc4R5260 as it's the only block where it's used.

**Review by luhenry — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 5 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: good catch, thanks!

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 6 above)
Comment (luhenry) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5259]: That `ProcessScalar` Label is only ever jumped to if we're in the `UseRVV` block, so please move that `__ BIND(ProcessScalar)` to L5256, inside the `if (UseRVV) { ... }` block.

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5259]: I see, You're right!

**Comment by camel-cdr** [comment]
@Hamlin-Li Hi, we looked at RVV base64 encode/decode for another project before, however there wasn't one implementation that obviously was best across the different hardware: https://github.com/WojciechMula/base64simd/issues/9 (see issue for benchmark, and repo for code)

I think we currently can't tell how, the complex load/stores will perform on future hardware. Segmented load/stores for example are quite fast on the current in-order RVV 1.0 boards, however it's very slow on the ooo C910, and XiangShan (current master, may change) cores (SiFive P670 LLVM-MCA indicates that it might also be slow on that core). I'm not sure if that is because they are ooo and that gives you additional constraints, but I wouldn't rely on it just yet.

I think the safest bet for encode would be for now "RISC-V RVV (LMUL=1)" ([`encode`](https://github.com/WojciechMula/base64simd/blob/master/encode/encode.rvv.cpp#L60C14-L60C20) + [`lookup_pshufb_improved`](https://github.com/WojciechMula/base64simd/blob/master/encode/lookup.rvv.cpp#L7)), as this only uses instructions with predictable performance, except for LMUL=1 `vrgather.vv`, which I think will need to be fast on any application class core. (See x86 equivalent vperm*)

For decode, I'm not really happy with any implementation. Yours uses multiple `vluxei8` + `vlsege4` + `vssege3`, the others from base64simd use LMUL=8 `vrgather.vv`, which will take `LMUL^2=8^2=64` times the amount of cycles a LMUL=1 `vrgather.vv` takes (on sane implementations, [see my reasoning](https://gitlab.com/riseproject/riscv-optimization-guide/-/issues/1#note_1977583125)). As I said, I'm fairly certain LMUL=1 `vrgather.vv` will have to be relatively fast, so if I had to choose, I'd prefer [my implementation](https://godbolt.org/z/7qc1xhMao) that uses LMUL=1 `vrgather.vv`s +  `vlsege4` + `vssege3`, but using `vsseg*` is not ideal. (Note that gcc currently chokes on the register allocation, so you should use clang for now)

**Comment by Hamlin-Li** [comment]
Thanks a lot for sharing the information.

> @Hamlin-Li Hi, we looked at RVV base64 encode/decode for another project before, however there wasn't one implementation that obviously was best across the different hardware: [WojciechMula/base64simd#9](https://github.com/WojciechMula/base64simd/issues/9) (see issue for benchmark, and repo for code)

Agree, I think your observation is right.

> I think we currently can't tell how, the complex load/stores will perform on future hardware. Segmented load/stores for example are quite fast on the current in-order RVV 1.0 boards, however it's very slow on the ooo C910, and XiangShan (current master, may change) cores (SiFive P670 LLVM-MCA indicates that it might also be slow on that core). I'm not sure if that is because they are ooo and that gives you additional constraints, but I wouldn't rely on it just yet.

I don't know how that (`it's very slow on the ooo`) happens and currently I don't have these types of machine. And it's bit strange that they are very slow with those instructions, could it be that they are not fully optimized for those instructions on these machines?

> I think the safest bet for encode would be for now "RISC-V RVV (LMUL=1)" ([`encode`](https://github.com/WojciechMula/base64simd/blob/master/encode/encode.rvv.cpp#L60C14-L60C20) + [`lookup_pshufb_improved`](https://github.com/WojciechMula/base64simd/blob/master/encode/lookup.rvv.cpp#L7)), as this only uses instructions with predictable performance, except for LMUL=1 `vrgather.vv`, which I think will need to be fast on any application class core. (See x86 equivalent vperm*)

My current tests on k230 shows that m2+m1+scalar bring the best performance on all size values, I'd like to see test data on other hardwares if someone can help test and get the data.
And, for current implementation it's easy to adjust lmul value in the algorithm. So I'm flexiable to either lmul value based the test data.

**Comment by Hamlin-Li** [comment]
> For decode, I'm not really happy with any implementation. Yours uses multiple `vluxei8` + `vlsege4` + `vssege3`, the others from base64simd use LMUL=8 `vrgather.vv`, which will take `LMUL^2=8^2=64` times the amount of cycles a LMUL=1 `vrgather.vv` takes (on sane implementations, [see my reasoning](https://gitlab.com/riseproject/riscv-optimization-guide/-/issues/1#note_1977583125)). As I said, I'm fairly certain LMUL=1 `vrgather.vv` will have to be relatively fast, so if I had to choose, I'd prefer [my implementation](https://godbolt.org/z/hrs61x9aP) that uses LMUL=1 `vrgather.vv`s + `vlsege4` + `vssege3`, but using `vsseg*` is not ideal. (Note that gcc currently chokes on the register allocation, so you should use clang for now)

I import [your implementation](https://godbolt.org/z/hrs61x9aP) into jdk, but compared to my current decode implementation, it brings much regression.
Let's discuss about decode in https://github.com/openjdk/jdk/pull/20026.

**Review by luhenry — APPROVED** [review]

**Comment by RealFYang** [comment]
Hi, will take a look. BTW: Have you resolved the performance issue of base64 decode instrinsic?

**Comment by Hamlin-Li** [comment]
> Hi, will take a look. 

Thanks!

> BTW: Have you resolved the performance issue of base64 decode instrinsic?

It only bring regression in MIME cases, I did some investigation, but not found root cause yet.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 7]
```diff
@@ -5103,6 +5103,223 @@ class StubGenerator: public StubCodeGenerator {
...
+
+    Register src = c_rarg0;
+    Register soff = c_rarg1;
+    Register send = c_rarg2;
+    Register dst = c_rarg3;
+    Register doff = c_rarg4;
+    Register isURL = c_rarg5;
+
+    Register codec = c_rarg6;
+    Register length = c_rarg7; // total length of src data in bytes
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5201]: Seems better to keep these aligned: 
```
    Register src   = c_rarg0;
    Register soff  = c_rarg1;
    Register send  = c_rarg2;
    Register dst   = c_rarg3;
    Register doff  = c_rarg4;
    Register isURL = c_rarg5;

    Register codec  = c_rarg6;
    Register length = c_rarg7; // total length of src data in bytes
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 8]
```diff
@@ -5103,6 +5103,223 @@ class StubGenerator: public StubCodeGenerator {
...
+
+    Label ProcessData, Exit, Greater;
+
+    RegSet saved_regs;
+    __ push_reg(saved_regs, sp);
+
+    // length should be multiple of 3
+    __ sub(length, send, soff);
+    __ mv(t0, 3);
+    __ div(length, length, t0);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5211]: Could this `div` be avoided? Maybe we could simply check and update `length` at the byte granunarity.

**Review by RealFYang — COMMENTED** [review]
Some minor comments.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 8 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: fixed.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 9]
```diff
@@ -5103,6 +5103,223 @@ class StubGenerator: public StubCodeGenerator {
...
+    Register doff = c_rarg4;
+    Register isURL = c_rarg5;
+
+    Register codec = c_rarg6;
+    Register length = c_rarg7; // total length of src data in bytes
+
+    Label ProcessData, Exit, Greater;
+
+    RegSet saved_regs;
+    __ push_reg(saved_regs, sp);
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: good catch, fixed.

**Comment by Hamlin-Li** [comment]
Hey, Is anyone available to take another look? Thanks!

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 10]
```diff
@@ -5103,6 +5103,226 @@ class StubGenerator: public StubCodeGenerator {
...
+    }
+
+    // scalar version
+    {
+      Register byte1 = soff, byte0 = send, byte2 = doff;
+      Register combined24Bits = isURL;
+      Register step = x28;
+
+      __ mv(step, 3);
+      __ blt(length, step, Exit);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5265]: Should we simply do `__ beqz(length, Exit)` instead here? That seems more readable to me.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 11]
```diff
@@ -5103,6 +5103,226 @@ class StubGenerator: public StubCodeGenerator {
...
+        // store 4 bytes encoded data
+        __ sb(byte0, Address(dst, 0));
+        __ sb(byte1, Address(dst, 1));
+        __ sb(byte2, Address(dst, 2));
+        __ sb(combined24Bits, Address(dst, 3));
+
+        __ sub(length, length, 3);
+        __ addi(dst, dst, 4);
+        // loop back
+        __ bge(length, step, ScalarLoop);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5314]: Similar here. Maybe `__ bnez(length, ScalarLoop)`?

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 12]
```diff
@@ -5103,6 +5103,226 @@ class StubGenerator: public StubCodeGenerator {
...
+
+      base64_vector_encode_round(src, dst, codec,
+                    size, stepSrcM2, stepDst,
+                    v2, v4, v6,         // inputs
+                    v8, v10, v12, v14,  // indexes
+                    v16, v18, v20, v22, // outputs
+                    Assembler::m2);
+
+      __ sub(length, length, stepSrcM2);
+      __ j(ProcessM2);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:5243]: Why not move the preceding `__ blt(length, stepSrcM1, ProcessScalar);` at line L5231 here and replace this direct jump to `ProcessM2`? That will save us one condition branch instruction in this `ProcessM2` loop. We only need to adjust the branch target of `__ blt(length, stepSrcM2, ProcessM1);` to here.

**Review by RealFYang — COMMENTED** [review]
Hi, Sorry for the late reply. Several more comments remain after a second look. Thanks.

**Comment by Hamlin-Li** [comment]
> Hi, Sorry for the late reply. Severa comments remain after a second look. Thanks.

No worry. :) Thanks for reviewing.

**Review by RealFYang — APPROVED** [review]
Updated change looks great. Thanks!

**Comment by Hamlin-Li** [comment]
@RealFYang @luhenry Thanks for your reviewing.

/integrate

**Pull request closed without merging**